### ISSUE TITLE
Wrong window content scaling on Qualcomm X Elite arm64 hardware

### ISSUE THREAD

**Issue by tobhe** [flutter/flutter#156816]
### Steps to reproduce

1. Boot Ubuntu on Lenovo Yoga Slim 7x or Dell XPS 13 9345 via https://people.canonical.com/~platform/images/ubuntu-concept/oracular-desktop-arm64+x1e.iso
2. Start any flutter application such as the graphical installer

This setup uses software rendering without GPU acceleration because of missing firmware at this point.

### Expected results

The window renders as expected, content fits the window frame and is readable.

### Actual results

Flutter seems to use a wrong scaling factor for window contents. Text appears to be cut-off and parts of the app are not visible at all.

### Code sample

```
$ flutter create -d linux foo
$ cd foo
$ flutter run
```

### Screenshots or Video

ubuntu graphical installer:
![image](https://github.com/user-attachments/assets/d6ba4ba4-400f-4a3c-97a8-5fe8344d0627)

example flutter app:
![image](https://github.com/user-attachments/assets/d674514b-3f39-417a-a590-f402c321a31c)

### Flutter Doctor output

```console
Initializing Flutter
Cloning into '/home/user/snap/flutter/common/flutter'...
warning: templates not found in /usr/share/git-core/templates
remote: Enumerating objects: 541482, done.
remote: Counting objects: 100% (456/456), done.
remote: Compressing objects: 100% (297/297), done.
remote: Total 541482 (delta 247), reused 281 (delta 131), pack-reused 541026 (from 1)
Receiving objects: 100% (541482/541482), 163.06 MiB | 12.08 MiB/s, done.
Resolving deltas: 100% (420133/420133), done.
Flutter initialized
Downloading Linux arm64 Dart SDK from Flutter engine 107ea8baa4f5b836f5393cb8585818881a302095...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  211M  100  211M    0     0  11.1M      0  0:00:18  0:00:18 --:--:-- 11.8M
Building flutter tool...
Resolving dependencies... (1.3s)
Downloading packages... (2.4s)
Got dependencies.
Flutter 3.27.0-1.0.pre.50 • channel master • https://github.com/flutter/flutter.git
Framework • revision 2b36701456 (34 minutes ago) • 2024-10-15 08:42:42 -0400
Engine • revision 107ea8baa4
Tools • Dart 3.6.0 (build 3.6.0-334.2.beta) • DevTools 2.40.1
Downloading Material fonts...                                      349ms
Downloading Gradle Wrapper...                                       31ms
Downloading package sky_engine...                                1,030ms
Downloading package flutter_gpu...                                 275ms
Downloading flutter_patched_sdk tools...                         1,067ms
Downloading flutter_patched_sdk_product tools...                 1,071ms
Downloading linux-arm64 tools...                                    3.5s
Downloading linux-arm64/font-subset tools...                     1,080ms
[✓] Flutter (Channel master, 3.27.0-1.0.pre.50, on Ubuntu 24.10 6.11.0-23-qcom-x1e, locale en_US.UTF-8)
    • Flutter version 3.27.0-1.0.pre.50 on channel master at /home/user/snap/flutter/common/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 2b36701456 (34 minutes ago), 2024-10-15 08:42:42 -0400
    • Engine revision 107ea8baa4
    • Dart version 3.6.0 (build 3.6.0-334.2.beta)
    • DevTools version 2.40.1

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/to/linux-android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✗] Chrome - develop for the web (Cannot find Chrome executable at google-chrome)
    ! Cannot find Chrome. Try setting CHROME_EXECUTABLE to a Chrome executable.

[✓] Linux toolchain - develop for Linux desktop
    • clang version 10.0.0-4ubuntu1
    • cmake version 3.16.3
    • ninja version 1.10.0
    • pkg-config version 0.29.1

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/to/linux-android-setup for detailed instructions).

[✓] Connected device (1 available)
    • Linux (desktop) • linux • linux-arm64 • Ubuntu 24.10 6.11.0-23-qcom-x1e

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 3 categories.
```

**Comment by huycozy** [comment]
Hi @tobhe 
Did you install Flutter via snap? Does the issue persist if you install Flutter manually? https://docs.flutter.dev/get-started/install/linux/desktop#install-the-flutter-sdk

**Comment by tobhe** [comment]
I did some more research and I think this might indeed be a snap issue. Will continue debugging, @robert-ancell has been helping me figure out what's going on. Some things I learnt so far:
- This is not specific to my device but happens on all devices with high enough resolution for gnome to default to 200% display scaling
- Manually setting scaling to 200% on non-affected machines causes the same issue
- Whatever flutter normally does to learn the scaling factor doesn't seem to work here and it defaults to 100%.
- I can't see any related apparmor warnings in my log

I also reproduced the issue with the example app, but since the flutter I used was the snap version `flutter run` also executes in snap context/confinement.
